app/src/main/java/com/example/gameworkflow/TimeMachineWorkflow.kt

```diff
@@ -0,0 +1,85 @@
+package com.example.gameworkflow
+
+import com.example.gameworkflow.TimeMachineWorkflow.State
+import com.squareup.workflow.*
+import com.squareup.workflow.WorkflowAction.Mutator
+
+/**
+ * Workflow that records its child's renderings and allows you to travel back in time.
+ */
+class TimeMachineWorkflow(
+    private val child: Workflow<Unit, Nothing, Any>,
+    private val shakeWorker: ShakeWorker
+) : StatefulWorkflow<Unit, State, Nothing, Any>() {
+
+    data class State(
+        val replayPoint: Int = PRESENT,
+        val history: MutableList<Any> = mutableListOf()
```

Comment:
This is the only part of this I don't like. Mutable things inside of state is a big no-no, but since this workflow is kind of meta it's the simplest way to do it.

Reply:
Seems like it wouldn't be hard to make history immutable, but I'm not going to lose any sleep over this.

Reply:
Maybe put a TODO comment in, to discourage imitators?